_pages/dev/ssr/server-side-rendering-contextual-logging.md

```diff
@@ -0,0 +1,244 @@
+  }
+
+  [...]
+}
+```
+
+An important aspect of our solution is its extensibility. We have designed our solution to be compatible with enterprise logging libraries used for logging in platform-server such as [Winston](https://www.npmjs.com/package/winston) or [Pino](https://www.npmjs.com/package/pino), making it easy to add their features to our system.
+
+The following is an example of how to implement the custom logger with the `Pino` library:
+
+```ts
+import { DefaultExpressServerLogger, ExpressServerLoggerContext } from '@spartacus/setup/ssr';
+import { pino } from 'pino';
+
+class CustomLogger extends DefaultExpressServerLogger {
+  private logger = pino({ customLevels: { log: 15 }, level: 'log' }); // as there is no `log` level in pino, we need to define it here;
+
+  [...]
+
+  // custom implementation of the `log` method
+  log(message: string, context: ExpressServerLoggerContext): void {
+    // custom log level is used here
+    this.logger.log(this.handleContext(context), message);
+  }
+
+  warn(message: string, context: ExpressServerLoggerContext): void {
+    //built-in log level is used here
+    this.logger.warn(this.handleContext(context), message);
+  }
+
+  [...]
+
+  // method to handle the context before it is passed to the logger
+  protected handleContext(context: ExpressServerLoggerContext) {
+    const { request, ...restOfContext } = context;
+    // `mapRequest` is used to map the request to the log message. This is a method provided in the `DefaultExpressServerLogger` class.
+    // NOTE: for some cases request object is not passed to the context, so we need to check if it exists
+    const mappedRequest = request ? this.mapRequest(request) : undefined;
+    return { ...restOfContext, request: mappedRequest };
+  }
+}
+```
+This is an example of how logs created by `pino` library look like:
+```
+{"level":15,"time":1687528211674,"pid":40850,"hostname":"SHMABCD1234567","options":{"concurrency":10,"timeout":3000,"forcedSsrTimeout":60000,"maxRenderTime":300000,"reuseCurrentRendering":true,"debug":false,"logger":"CustomLogger"},"msg":"[spartacus] SSR optimization engine initialized"}
+
+{"level":15,"time":1687528382249,"pid":40850,"hostname":"SHMABCD1234567","request":{"url":"/electronics-spa/en/USD/","uuid":"446d5d59-0111-471e-b6f3-883617a9aaba","timeReceived":"2023-06-23T13:53:02.247Z"},"msg":"Rendering started (/electronics-spa/en/USD/)"}
+```
+For more information, see: https://www.npmjs.com/package/pino
+
+Note: This is only an example of how to use a third-party logger library. Do not treat this as a recommendation of the particular library.
+
+**Caution:** The `DefaultExpressServerLogger` doesn't include any sensitive data in the logs by default. However, when using a custom logger, developers must exercise caution and be mindful of the data being provided. It's their responsibility to ensure that no sensitive information is being used or exposed.
+
+## Using the LoggerService
```

Comment:
We should strongly emphasise in this doc, that customers should use `LoggerService` instead of the global `console` if they want to automatically add request's context to their custom logs. And also any 3rd party libraries that print to the global console, that they use in their project, should also call LoggerService rather than the global `console` object.

--
Btw. We could also extract doc for LoggerService deserves into a separate standalone document. It can be mentioned in this "SSR contextual logs" doc (and vice-versa).

Reply:
As we discussed, for now, let's keep this information in the section about SSR. We can discuss moving the `LoggerService` description later, e.g. before the next major release.